src/core/lib/iomgr/timer_generic.c

```diff
@@ -178,7 +191,7 @@ void grpc_timer_init(grpc_exec_ctx *exec_ctx, grpc_timer *timer,
   GPR_ASSERT(deadline.clock_type == g_clock_type);
   GPR_ASSERT(now.clock_type == g_clock_type);
   timer->closure = closure;
-  timer->deadline = deadline;
+  timer->deadline = timespec_to_atm(deadline);
```

Comment:
So timespec_to_atm seems to always round down, but the timer specification is that we won't fire before the specified deadline. In this case, we'll fire up to .999 ms before the specified deadline and that may cause some issues. I think it would be fine to round up in timespec_to_atm without violating our expectations of timers, though this change does limit the use of very high-frequency timers (which were not guaranteed to be very precise anyway)